### PR TITLE
WIP: Bbox plugin that works with RavenML core (global-dataset-plugin-changes branch)

### PR DESCRIPTION
**Pull request by AbhiDhir**
This PR adds a functioning bbox plugin for raven core that can be used for dataset creation. Currently, the interfaces for the plugin are defined as:

Usage:

- `ravenml data create -c <config file> tf-bbox`

Other than that, there are a few specific spots in the code that should be looked at which I will comment on.

EDITS:

- RMLCache now in use
- DatasetWriter class from core now in use
- Removed object_detection dependencies in tf export function
- Makes download and filter overridable by plugin

### CODE REVIEW

File: rmldatatfbbox/rmldatatfbbox/core.py [hunk 1]
```diff
@@ -5,77 +5,158 @@
...
+@pass_create
+@click.pass_context
+def create(ctx: click.Context, create: CreateInput):
+
+    cli_spinner("Importing TensorFlow...", _import_od)
+
+    config = create.config.get("plugin")
+
+    # Optional Information from interface (just contains plugin)
+    # metadata = create.plugin_metadata
```
Comment (AbhiDhir) [inline rmldatatfbbox/rmldatatfbbox/core.py]: This metadata field only contains the plugin's name so not really useful right now

File: rmldatatfbbox/rmldatatfbbox/core.py [hunk 2]
```diff
@@ -5,77 +5,158 @@
...
+        "image_type_3": ".jpeg",
+        "metadata": ".json",
+        "labels": ".csv",
+        "PASCAL_VOC_labels": ".xml"
+    }
+
+    related_data_prefixes = {
+        'images': 'image_',
+        'labels': 'bboxLabels_'
+    }
```
Comment (AbhiDhir) [inline rmldatatfbbox/rmldatatfbbox/core.py]: These two lists were taken out of the BBox classes, it may be cleaner to create json/yaml files for them as a standard

File: rmldatatfbbox/rmldatatfbbox/core.py [hunk 3]
```diff
@@ -5,77 +5,158 @@
...
+    cli_spinner("Writing out dataset locally...", write_dataset,list(labeled_images.values()),
+                                                                     custom_dataset_name=dataset_name,
+                                                                     num_folds=int(k_folds_specified),
+                                                                     test_percent=test_percent,
+                                                                     out_dir=base_dir,
+                                                                     associated_files=associated_files,
+                                                                     related_data_prefixes=related_data_prefixes,
+                                                                     label_to_int_dict=label_to_int_dict,
+                                                                     export_function=export_as_TFExample)
+
```
Comment (AbhiDhir) [inline rmldatatfbbox/rmldatatfbbox/core.py:51]: The last parameter "export_function" is passing the specific function that writes the tfrecords, could be a cleaner solution (extracting the function that uses it from raven core back to plugin)

File: rmldatatfbbox/rmldatatfbbox/utils/helpers.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    with open(data_dir / f"meta_{image_id}.json", "r") as f:
+        meta = json.load(f)
+
+    label_boxes = []
+    for label, b in meta['bboxes'].items():
+        label_boxes.append({"label": label, "xmin": b['xmin'], "xmax": b['xmax'], "ymin": b['ymin'], "ymax": b['ymax']})
+        if label not in label_to_int_dict.keys():
+            label_to_int_dict[label] = len(label_to_int_dict)
+    
+    bboxes = {"image_id": image_id, "image_filepath": image_filepath, "image_type": image_type, "label_boxes": label_boxes, "xdim": xdim, "ydim": ydim}
```
Comment (AbhiDhir) [inline rmldatatfbbox/rmldatatfbbox/utils/helpers.py]: This is where the classes were used before, and the label_to_int_dict is now a parameter that is passed throughout the dataset creation process instead of being stored in a class

File: rmldatatfbbox/sample_configs/bbox_all_fields.yaml [hunk 5]
```diff
@@ -0,0 +1,19 @@
+# This sample config contains all fields supported by ravenML core and the bbox plugin.
+# Plugin specific configuration is located in the plugin field.
+
+imageset: # Can provide a list of imagesets if needed
+  - jigsaw_tester
```
Comment (AbhiDhir) [inline rmldatatfbbox/sample_configs/bbox_all_fields.yaml:7]: If one wants to take images from multiple imagesets it is supported by just adding more lines with the prefixes

File: rmldatatfbbox/sample_configs/bbox_all_fields.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+overwrite_local: True
+dataset_path: '~/Desktop/test'
+# filter: True
```
Comment (AbhiDhir) [inline rmldatatfbbox/sample_configs/bbox_all_fields.yaml]: Filtering is supported but an optional field that is default False

File: rmldatatfbbox/sample_configs/bbox_all_fields.yaml [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+upload: False
+delete_local: True
+metadata:
+    created_by: Abhi Dhir
+    comments: test
+plugin:
+    verbose: True
+    kfolds: 5
+    dataset_name: bbox_test_dataset_abhi
+    test_percent: .2 # optional
```
Comment (AbhiDhir) [inline rmldatatfbbox/sample_configs/bbox_all_fields.yaml]: test_percent is something I added since it might be useful to modify from the default (.2) but is also optional (won't prompt if not provided)

File: rmldatatfbbox/rmldatatfbbox/core.py [hunk 8]
```diff
@@ -5,77 +5,158 @@
...
-
-### OPTIONS ###
-
+import sys
+import io
+import importlib
+import shutil
+from pathlib import Path
+import tensorflow as tf
+from rmldatatfbbox.utils.helpers import (construct_all, write_label_map, export_as_TFExample)
```
Comment (kvablack) [inline rmldatatfbbox/rmldatatfbbox/core.py]:
```suggestion
from rmldatatfbbox.utils.helpers import construct_all, write_label_map, export_as_TFExample
```
nit: remove redundant parentheses

File: rmldatatfbbox/rmldatatfbbox/core.py [hunk 9]
```diff
@@ -5,77 +5,158 @@
...
-#     #
-#     # import tensorflow as tf
-#     # from object_detection import model_hparams
-#     # from object_detection import model_lib
-#     _dynamic_import('tensorflow', 'tf')
-#     _dynamic_import('object_detection.model_hparams', 'model_hparams')
-#     _dynamic_import('object_detection.model_lib', 'model_lib')
-#     _dynamic_import('object_detection.exporter', 'exporter')
-#     _dynamic_import('object_detection.protos', 'pipeline_pb2', asfunction=True)
+    k_folds_specified = config.get("kfolds") if config.get("kfolds") else user_input(message="How many folds would you like the dataset to have?",default="5")
```
Comment (kvablack) [inline rmldatatfbbox/rmldatatfbbox/core.py]: Should default to no k-fold (probably don't even need to do user input, just assume no k-fold unless they specifically request it)

File: rmldatatfbbox/rmldatatfbbox/core.py [hunk 10]
```diff
@@ -5,77 +5,158 @@
...
+
+    test_percent = config.get("test_percent") if config.get("test_percent") else .2
     
-#     # now restore stdout function
-#     sys.stdout = sys.__stdout__
-#     sys.stderr = sys.__stderr__
+    imageset_data = create.config.get("metadata").get("imageset_data")
+    labeled_images, label_to_int_dict = construct_all(imageset_data.get("image_ids"), imageset_data.get("temp_dir"))
```
Comment (kvablack) [inline rmldatatfbbox/rmldatatfbbox/core.py]:
```suggestion
    imageset_data = create.config.["metadata"]["imageset_data"]
    labeled_images, label_to_int_dict = construct_all(imageset_data["image_ids"], imageset_data["temp_dir"])
```
nit: no need to use `.get` unless you specifically want it to default to `None` (same with other instances of this pattern below)

File: rmldatatfbbox/rmldatatfbbox/core.py (hunk 9 above)
Comment (kvablack) [reply rmldatatfbbox/rmldatatfbbox/core.py]: Actually, I see that the k-fold logic is still commented out in `write_dataset` in ravenML core. There will obviously have to be more broad interface changes to make k-fold optional.

File: rmldatatfbbox/rmldatatfbbox/core.py [hunk 11]
```diff
@@ -5,77 +5,158 @@
...
+        "labels": ".csv",
+        "PASCAL_VOC_labels": ".xml"
+    }
+
+    related_data_prefixes = {
+        'images': 'image_',
+        'labels': 'bboxLabels_'
+    }
+
+    cli_spinner("Writing out dataset locally...", write_dataset,list(labeled_images.values()),
```
Comment (kvablack) [inline rmldatatfbbox/rmldatatfbbox/core.py]: Wrapping in `cli_spinner` should probably happen inside `write_dataset` in the core so the plugin author doesn't have to worry about it. Same with other functions that come from the core.

File: rmldatatfbbox/rmldatatfbbox/utils/helpers.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+    for image_id in image_ids:
+        labeled_images[image_id] = construct(image_id, temp_dir, label_to_int_dict)
+    
+    return labeled_images, label_to_int_dict
+
+def construct(image_id: str, temp_dir: Path, label_to_int_dict: dict):
+    if temp_dir is None:
+        cwd = Path.cwd()
+        data_dir = cwd / 'data'
```
Comment (kvablack) [inline rmldatatfbbox/rmldatatfbbox/utils/helpers.py]: This is mainly a change to the core, but we should get rid of any use of the CWD (i.e. `cwd/data` and `cwd/dataset`). These should instead use the RavenML cache in `~/.ravenML/imagesets` and `~/.ravenML/datasets`.

File: rmldatatfbbox/rmldatatfbbox/utils/helpers.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+    else:
+        data_dir = temp_dir
+
+    image_filepath = None
+    image_type = None
+    file_extensions = [".png", ".jpg", ".jpeg"]
+    for extension in file_extensions:
+        if os.path.exists(data_dir / f'image_{image_id}{extension}'):
```
Comment (kvablack) [inline rmldatatfbbox/rmldatatfbbox/utils/helpers.py]: Again, this is part of the more broad discussion about the interface between the core and the plugins. But I'm thinking instead of the core doing the `image_ids` discovery and therefore locking the filenames into a very specific format, it makes more sense to let each plugin glob for whatever data it needs in whatever way it wants and then pass lists of paths back to the core.

File: rmldatatfbbox/rmldatatfbbox/utils/helpers.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+        for bounding_box in object["label_boxes"]:
+            xmins.append(bounding_box["xmin"] / image_width)
+            xmaxs.append(bounding_box["xmax"] / image_width)
+            ymins.append(bounding_box["ymin"] / image_height)
+            ymaxs.append(bounding_box["ymax"] / image_height)
+            classes_text.append(bounding_box["label"].encode('utf8'))
+            classes.append(label_to_int_dict[bounding_box["label"]])
+
+        tf_example = tf.train.Example(features=tf.train.Features(feature={
+            'image/height': dataset_util.int64_feature(image_height),
```
Comment (kvablack) [inline rmldatatfbbox/rmldatatfbbox/utils/helpers.py]: We should break our dependency with object-detection. It makes absolutely no sense, because if you look at what `dataset_util` is actually doing, it's basically just aliases to easily accessible Tensorflow functions :facepalm:

**Review by kvablack — CHANGES_REQUESTED** [review]
This is an awesome start! The way you've split the functionality between the plugin and the core is already beginning to strike a decent balance between flexibility and usability for plugin authors. It looks relatively clean considering how much Jigsaw code you were able to reuse; perhaps we won't need as many sweeping changes as I originally thought, which is a really good thing.

Most of my suggestions are nitpicks, since any bigger changes to the interface will have to involve the core. You should open a corresponding pull request to ravenML so we can start having a discussion there.

File: rmldatatfbbox/rmldatatfbbox/core.py (hunk 8 above)
Comment (kvablack) [reply rmldatatfbbox/rmldatatfbbox/core.py]: Oh, also, the `utils` directory needs an `__init__.py` in order to get recognized by `find_packages()`.

File: rmldatatfbbox/rmldatatfbbox/core.py (hunk 3 above)
Comment (kvablack) [reply rmldatatfbbox/rmldatatfbbox/core.py:51]: Just an idea I had: what if instead of these single ugly functions with like 10 arguments, we went partially back to the object-oriented paradigm by creating something like a DatasetWriter class in the core. It could handle any universal stuff (e.g. dataset name, user, comments, training type, maybe even filters and transforms) so the plugin didn't have to worry about it at all. It could also have default/convenience implementations for some more common functionality, like the `label_to_int_dict`. Plus, it could take advantage of `self` to store stuff like the `label_to_int_dict` and all the metadata without having to pass them around in big dictionaries.

Then, plugins could subclass DatasetWriter and override as much functionality as they wanted. Most plugins would probably only need to override methods akin to `construct_all` and `export_as_TFExample`, and the rest would be handled for them. However, if a plugin really needed to, it could override as much functionality as it wanted, even `write_dataset` if it wanted to completely customize the structure of the output.

Basically, this is similar to what old Jigsaw did except without the weird class-level metadata and more flexible. I think this could be a bit cleaner while keeping even more unnecessary details out of the plugin. It could also be a starting point to split up big monolithic functions like `write_dataset` so that the plugin can override specific bits of functionality in an even more modular fashion.

File: rmldatatfbbox/rmldatatfbbox/core.py [hunk 15]
```diff
@@ -5,77 +5,158 @@
...
+from rmldatatfbbox.utils.helpers import (construct_all, write_label_map, export_as_TFExample)
+from ravenml.utils.question import user_input, cli_spinner
+from ravenml.data.helpers import default_filter_and_load
+from ravenml.data.write_dataset import write_dataset, write_metadata
+from ravenml.data.options import pass_create
+from ravenml.data.interfaces import CreateInput, CreateOutput
 
 ### COMMANDS ###
-@click.command(help='TensorFlow Object Detection with bounding boxes.')
+@click.group(help='Dataset creation with bounding box info')
```
Comment (cSchubes) [inline rmldatatfbbox/rmldatatfbbox/core.py]: This should specifically say TensorFlow, since the output dataset is only for use with TF

File: rmldatatfbbox/rmldatatfbbox/core.py (hunk 1 above)
Comment (cSchubes) [reply rmldatatfbbox/rmldatatfbbox/core.py]: See the `__init__` in `TrainInput` (https://github.com/autognc/ravenML/blob/master/ravenml/train/interfaces.py#L44) for things to add to metadata. The metadata should be created by ravenML core and passed to the plugin for the plugin to add anything else it wants. Things like author name, dataset type, etc should be set by the core

File: rmldatatfbbox/rmldatatfbbox/core.py [hunk 16]
```diff
@@ -5,77 +5,158 @@
...
-#     # create a text trap and redirect stdout
-#     # to suppress printed warnings from object detection and tf
-#     text_trap = io.StringIO()
-#     sys.stdout = text_trap
-#     sys.stderr = text_trap
+    # Transformation functionality is not currently supported, thus the empty array
+    transform_metadata = []
+
+    # Variable type validation is currently not supported
+    dataset_name = config.get("dataset_name") if config.get("dataset_name") else user_input(message="What would you like to name this dataset?")
```
Comment (cSchubes) [inline rmldatatfbbox/rmldatatfbbox/core.py]: This is functionality that should be placed into the `__init__` for the `CreateInput`, see same lines as I referenced above (https://github.com/autognc/ravenML/blob/master/ravenml/train/interfaces.py#L44)

File: rmldatatfbbox/rmldatatfbbox/core.py (hunk 9 above)
Comment (cSchubes) [reply rmldatatfbbox/rmldatatfbbox/core.py]: I think having ability to do k fold is good, at least if we already have the code to do it. but yes it should be optional and unprompted, similar to the `ec2_policy` field used in the training plugin configs

File: rmldatatfbbox/rmldatatfbbox/core.py (hunk 2 above)
Comment (cSchubes) [reply rmldatatfbbox/rmldatatfbbox/core.py]: I think for now its ok to leave it directly in the code, unless these lists get extremely long

File: rmldatatfbbox/rmldatatfbbox/core.py (hunk 11 above)
Comment (cSchubes) [reply rmldatatfbbox/rmldatatfbbox/core.py]: agreed, can just use a private helper if needed

File: rmldatatfbbox/rmldatatfbbox/core.py (hunk 3 above)
Comment (cSchubes) [reply rmldatatfbbox/rmldatatfbbox/core.py:51]: class based could be good, nothing inherently wrong with OO stuff. I know this PR isn't for the core but I looked at `write_dataset` and we definitely need to make some changes to make this fully extensible, that function is still calling other functions that are specific to TF. We should extract out TF related functionality to a module in the core and let plugins manually specify use of those common tools.

File: rmldatatfbbox/rmldatatfbbox/utils/helpers.py (hunk 12 above)
Comment (cSchubes) [reply rmldatatfbbox/rmldatatfbbox/utils/helpers.py]: Believe me I know. I left use of CWD in Jigsaw because it was so integrated throughout I couldn't get rid of it easily... for the love of god and all that is holy, Python devs, please stop using CWD

File: rmldatatfbbox/rmldatatfbbox/utils/helpers.py (hunk 13 above)
Comment (cSchubes) [reply rmldatatfbbox/rmldatatfbbox/utils/helpers.py]: I'm confused, this code is all in the plugin, not the core. Am I missing something? looks like the plugin is doing its own discovery here

File: rmldatatfbbox/rmldatatfbbox/utils/helpers.py (hunk 14 above)
Comment (cSchubes) [reply rmldatatfbbox/rmldatatfbbox/utils/helpers.py]: ^ big bump, if we can just use TF that would be great. I always wondered why we needed OD for the datasets

File: rmldatatfbbox/sample_configs/bbox_all_fields.yaml (hunk 6 above)
Comment (cSchubes) [reply rmldatatfbbox/sample_configs/bbox_all_fields.yaml]: if you set filter to true does it result in prompted filtering? and where is the logic for the filtering located? (core vs. plugin)

File: rmldatatfbbox/sample_configs/bbox_all_fields.yaml [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+# filter: True
+upload: False
+delete_local: True
+metadata:
+    created_by: Abhi Dhir
+    comments: test
+plugin:
+    verbose: True
+    kfolds: 5
+    dataset_name: bbox_test_dataset_abhi
```
Comment (cSchubes) [inline rmldatatfbbox/sample_configs/bbox_all_fields.yaml]: `dataset_name` should not be in the plugin section as it is a base field required by all plugins. same may apply to `test_percent` if we decide that is basic enough to force onto every plugin, and possibly even `kfolds`

File: .gitignore [hunk 18]
```diff
@@ -115,6 +115,11 @@ venv.bak/
 # compiled protos
 *pb2.py
 
+#Json folder
```
Comment (cSchubes) [inline .gitignore]: you said its in progress so just placing this comment here so we don't forget since this will be a living PR.

we should stop using `cwd` and place locally created datasets into the RMLCache, then we can remove these gitignore paths

File: rmldatatfbbox/rmldatatfbbox/core.py [hunk 19]
```diff
@@ -5,77 +5,158 @@
...
 @click.pass_context
 def tf_bbox(ctx):
     pass
+
+@tf_bbox.command(help='Create a dataset.')
+@pass_create
+@click.pass_context
+def create(ctx: click.Context, create: CreateInput):
+
+    cli_spinner("Importing TensorFlow...", _import_od)
```
Comment (cSchubes) [inline rmldatatfbbox/rmldatatfbbox/core.py]: I'm glad to see this function getting used, however in the current setup it is useless since you're importing tensorflow directly on line 13 of `core.py`. Notice that in `import_od` the tensorflow library, along with other OD libraries, are imported. 

This raises the larger point of whether we want to try and keep using this pattern or not. Personally I think we should, because it's incredibly annoying to wait 2-3s for TF to be imported on every command you run with ravenML, even basic data exploration commands.

On the other hand, if even one plugin doesn't use this pattern, it defeats the purpose entirely. If we are truly open-sourcing this it could get hard to enforce the pattern. Personally I think we should try and require this of plugin developers, because I don't think it's very complicated to use `_dynamic_import` and it makes raven way more usable. But i can see both sides

File: rmldatatfbbox/sample_configs/bbox_all_fields.yaml (hunk 6 above)
Comment (AbhiDhir) [reply rmldatatfbbox/sample_configs/bbox_all_fields.yaml]: Yes, if filter is set to true, prompted filtering is the result and this happens in the core when initializing the create input.

File: rmldatatfbbox/rmldatatfbbox/utils/helpers.py (hunk 13 above)
Comment (kvablack) [reply rmldatatfbbox/rmldatatfbbox/utils/helpers.py]: Sorry, this isn't the most relevant bit of code to comment on. But right now the core is responsible for discovering all of the image IDs. Based on my brief look, it does this by explicitly searching for `meta_*.json` files and extracting the IDs from there (see `data/helpers.py -> load_metadata`). The interactive filtering happens on these IDs before any plugin code is run. The plugin then gets the list of IDs and uses those to re-infer all the filepaths, which is what you see above.

The plugin should be at least able to override this behavior. Just continuing to spitball based on my earlier object-oriented proposal: perhaps there could be a DatasetWriter method along the lines of `load_ids` or something. It would take the path to the data and be expected to return a dict with IDs as keys. DatasetWriter would do the interactive filtering on these IDs, then pass along the dict to further methods. The default implementation could be similar to the existing one, where the IDs are inferred from `meta_*.json` and the dict values are just `(metadata_path, image_path)` tuples. However, if a plugin wanted to, it could override this method to do ID discovery however it wanted and store any arbitrary amount of information as the dict values.

File: rmldatatfbbox/rmldatatfbbox/core.py (hunk 19 above)
Comment (AbhiDhir) [reply rmldatatfbbox/rmldatatfbbox/core.py]: I have removed it from the core.py file, but I can't remove it from the helpers.py file without it throwing an error. Is there something I am missing?

Comment (kvablack) [reply rmldatatfbbox/rmldatatfbbox/core.py]: Shouldn't have to worry about this if https://github.com/autognc/ravenML/pull/43 gets merged, so marking resolved for now.

File: rmldatatfbbox/rmldatatfbbox/core.py [hunk 20]
```diff
@@ -5,77 +5,121 @@
...
+from pathlib import Path
+from rmldatatfbbox.utils.helpers import BboxDatasetWriter
+from ravenml.utils.question import user_input, cli_spinner
+from ravenml.data.helpers import default_filter_and_load
+from ravenml.data.options import pass_create
+from ravenml.data.interfaces import CreateInput, CreateOutput
 
 ### COMMANDS ###
-@click.command(help='TensorFlow Object Detection with bounding boxes.')
+@click.group(help='TensorFlow Object Detection with bounding boxes.')
```
Comment (kvablack) [inline rmldatatfbbox/rmldatatfbbox/core.py]: nit, but if there's only one command, we should maybe consider forgoing the group. I find the `ravenml data create tf-bbox create` syntax a bit confusing compared to just `ravenml data create tf-bbox`.

Comment (AbhiDhir) [reply rmldatatfbbox/rmldatatfbbox/core.py]: Yeah, that was something I was thinking as well

File: rmldatatfbbox/sample_configs/bbox_all_fields.yaml (hunk 6 above)
Comment (cSchubes) [reply rmldatatfbbox/sample_configs/bbox_all_fields.yaml]: what do we think of exposing a bit more filter functionality through the config file? I think at the very least being able to tell it how many images you want without incurring all the prompts would be good. In that case it would evenly split those images between all the imagesets.

File: rmldatatfbbox/rmldatatfbbox/core.py (hunk 9 above)
Comment (cSchubes) [reply rmldatatfbbox/rmldatatfbbox/core.py]: if we are onboard with optional kfold as described here (https://github.com/autognc/ravenML/pull/44#discussion_r456215693) then let's change this to unprompted behavior only controllable via config file. You can use the presence of a "kfold" field to determine if its enabled, and the value of the field can be the # of folds.

Also, maybe good to add a check somewhere to ensure folds < # of images. Otherwise dataset creation will fail on the split logic.

File: rmldatatfbbox/rmldatatfbbox/utils/helpers.py (hunk 13 above)
Comment (cSchubes) [reply rmldatatfbbox/rmldatatfbbox/utils/helpers.py]: can you elaborate more on the use case for the generalized functionality @kevin3-black ? As long as we are using imagery generated by Starfish, we are certain each image will have a metadata file to hold labels and tags. Are you thinking of cases where image labels and/or tags may be stored in other file formats or naming schemes different from what we have defined in starfish?

File: rmldatatfbbox/rmldatatfbbox/core.py (hunk 20 above)
Comment (cSchubes) [reply rmldatatfbbox/rmldatatfbbox/core.py]: I've thought about this a lot and decided against this simplification for one reason:

While `ravenml data create tf-bbox create` is unwieldy compared to `ravenml data create tf-bbox`, it ensures a consistent interface across plugins that may have > 1 command. This was particularly apparent on the training side where multiple plugins already have > 1 command. I think it is more confusing to a user to jump between plugins defined as a single command vs. plugins defined as a command group than it is to just type one extra word.

Caveat to this in light of https://github.com/autognc/ravenML/pull/43: because of lazy loading, autocompletion (which we can and should add to ravenML) will no longer work for plugins. Thus maybe it is more convenient to just let plugins choose to be a group OR a single command, as users will not be able to autocomplete that pesky extra `create`.

File: rmldatatfbbox/rmldatatfbbox/core.py [hunk 21]
```diff
@@ -72,31 +44,17 @@ def create(ctx: click.Context, create: CreateInput):
...
 
-    dataset_path = base_dir / 'dataset' / dataset_name
+    datasetWriter.write_dataset(list(labeled_images.values()))
+    datasetWriter.write_metadata()
+    datasetWriter.write_additional_files()
 
-    return CreateOutput(dataset_path, dataset_name, create.config)
+    return CreateOutput(create)
 
 ### HELPERS ###
```
Comment (cSchubes) [inline rmldatatfbbox/rmldatatfbbox/core.py]: this should be removed now thanks to https://github.com/autognc/ravenML/pull/43

File: rmldatatfbbox/rmldatatfbbox/core.py [hunk 22]
```diff
@@ -72,31 +44,17 @@ def create(ctx: click.Context, create: CreateInput):
...
-
-    cli_spinner("Writing out metadata locally...", write_metadata,name=dataset_name,
-                                                                  user=create.config.get("metadata").get("created_by"),
-                                                                  comments=create.config.get("metadata").get("comments"),
-                                                                  training_type="Bounding_Box",
-                                                                  image_ids=imageset_data.get("image_ids"),
-                                                                  filters=imageset_data.get("filter_metadata"),
-                                                                  transforms=transform_metadata,
-                                                                  out_dir=base_dir)
+    datasetWriter = BboxDatasetWriter(create, associated_files=associated_files, related_data_prefixes=related_data_prefixes)
```
Comment (cSchubes) [inline rmldatatfbbox/rmldatatfbbox/core.py]: Since bbox is our baseline (and also one of the simpler cases), do we expect this pattern to happen often in `create` functions? It seems that thanks to the DatasetWriter this `create` logic is generic. Maybe the core can provide an implementation that takes in the DatasetWriter and does the logic as listed here. a plugin that just wants to do the basic thing can populate their `create` function with a single call to that imported function. Thoughts?

Ironically, this brings us closer to the original Jigsaw implementation and raises some questions as to whether a whole plugin is really necessary for each dataset type. I still say yes, as it provides flexibility for future, more complex datasets. But it is a good bit of boilerplate. I am planning on providing baseline training plugin code via either A) GitHub gist or B) directly in the `ravenML-training-plugins` repo so a user can just copy and paste that and avoid most of the boilerplate. Would be a good idea to do the same for dataset plugins, including a baseline implementation (or at least skeleton with TODOs) of the DatasetWriter

File: rmldatatfbbox/rmldatatfbbox/utils/helpers.py (hunk 13 above)
Comment (cSchubes) [reply rmldatatfbbox/rmldatatfbbox/utils/helpers.py]: resolved from 7/22 meeting discussion

**Review by kvablack — APPROVED** [review]
